mining: check witness commitment in submitBlock

Sjors opened the pull request:
When an IPC client requests a new block template via the Mining interface, we hold on to its `CBlock`. That way when they call `submitSolution()` we can modify it in place, rather than having to reconstruct the full block like the `submitblock` RPC does.

Before this commit however we forgot to invalidate `m_checked_witness_commitment`, which we should since the client brings a new coinbase.

This would cause us to accept an invalid chaintip.

Fix this and add a test to confirm that we now reject such a block. As a sanity check, we add a second node to the test and confirm that will accept our mined block.

As first noticed in #33374 the IPC code takes the coinbase as provided, unlike the `submitblock` RPC which calls `UpdateUncommittedBlockStructures()` and adds witness commitment to the coinbase if it was missing.

Although that could have been an alternative fix, we instead document that IPC clients are expected to provide the full coinbase including witness commitment.

Patch to produce the original issue:

```diff
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index b988e28a3f..28e9048a4d 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -450,15 +450,10 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
     }
     block.nVersion = version;
     block.nTime = timestamp;
     block.nNonce = nonce;
     block.hashMerkleRoot = BlockMerkleRoot(block);
-
-    // Reset cached checks
-    block.m_checked_witness_commitment = false;
-    block.m_checked_merkle_root = false;
-    block.fChecked = false;
 }

 std::unique_ptr<CBlockTemplate> WaitAndCreateNewBlock(ChainstateManager& chainman,
                                                       KernelNotifications& kernel_notifications,
                                                       CTxMemPool* mempool,
diff --git a/test/functional/interface_ipc.py b/test/functional/interface_ipc.py
index cce56e3294..bf1b7048ab 100755
--- a/test/functional/interface_ipc.py
+++ b/test/functional/interface_ipc.py
@@ -216,22 +216,22 @@ class IPCInterfaceTest(BitcoinTestFramework):
             assert_equal(res.result, True)

             # The remote template block will be mutated, capture the original:
             remote_block_before = await self.parse_and_deserialize_block(template, ctx)

-            self.log.debug("Submitted coinbase must include witness")
+            self.log.debug("Submitted coinbase with missing witness is accepted")
             assert_not_equal(coinbase.serialize_without_witness().hex(), coinbase.serialize().hex())
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize_without_witness())
-            assert_equal(res.result, False)
+            assert_equal(res.result, True)

             self.log.debug("Even a rejected submitBlock() mutates the template's block")
             # Can be used by clients to download and inspect the (rejected)
             # reconstructed block.
             remote_block_after = await self.parse_and_deserialize_block(template, ctx)
             assert_not_equal(remote_block_before.serialize().hex(), remote_block_after.serialize().hex())

-            self.log.debug("Submit again, with the witness")
+            self.log.debug("Submit again, with the witness - does not replace the invalid block")
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize())
             assert_equal(res.result, True)

             self.log.debug("Block should propagate")
             assert_equal(self.nodes[1].getchaintips()[0]["height"], current_block_height + 1)
```

fanquake commented:
https://github.com/bitcoin/bitcoin/actions/runs/18939943748/job/54075967306?pr=33745#step:9:2351:
```bash

test/miner_tests.cpp(70): Entering test suite "miner_tests"
test/miner_tests.cpp(686): Entering test case "CreateNewBlock_validity"
<snip>
 2025-10-30T12:08:21.221330Z [test] [validationinterface.cpp:218] [void ValidationSignals::BlockConnected(ChainstateRole, const std::shared_ptr<const CBlock> &, const CBlockIndex *)] [validation] Enqueuing BlockConnected: block hash=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f block height=0
2025-10-30T12:08:21.221451Z [scheduler] [validationinterface.cpp:218] [auto ValidationSignals::BlockConnected(ChainstateRole, const std::shared_ptr<const CBlock> &, const CBlockIndex *)::(anonymous class)::operator()() const] [validation] BlockConnected: block hash=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60node/miner.cpp:187 CreateNewBlock: Assertion `pblock->m_checked_witness_commitment' failed.
```

Sjors commented on src/interfaces/mining.h:
```diff
@@ -54,7 +54,15 @@ class BlockTemplate
+     * @note unlike the submitblock RPC, this method does NOT add the
+     *       coinbase witness automatically.
      *
      * @returns if the block was processed, independent of block validity
```
It behaves the same way as `ProcessNewBlock`, which has the following confusing documentation:

> This only returns after the best known valid
> block is made active. Note that it does not, however, guarantee that the
> specific block passed to it has been checked for validity!

davidgumberg replied:
Maybe the comment here could be changed to something like:

```suggestion
     * @returns if the block was processed, does not necessarily indicate validity.
```

TheCharlatan commented on test/functional/interface_ipc.py:
```diff
@@ -204,13 +208,32 @@ async def async_routine():
+
+            self.log.debug("Submitted coinbase must include witness")
+            assert_not_equal(coinbase.serialize_without_witness().hex(), coinbase.serialize().hex())
+            res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize_without_witness())
```
In commit 504acfcd5d1ed4d77f7fd4f5dc57e66d6718d3d4:

If I apply this test independently before the miner.cpp code changes, it still passes, so not sure what we are testing here. Wouldn't this only be relevant if we'd run checkBlock on the template's own block? Afaict we only do checkBlock on blocks we re-create here.

davidgumberg replied:
Strange, for me this test fails as expected with the following diff applied to this branch:

```diff
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index b988e28a3f..485984e437 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -454,9 +454,9 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
     block.hashMerkleRoot = BlockMerkleRoot(block);

     // Reset cached checks
-    block.m_checked_witness_commitment = false;
-    block.m_checked_merkle_root = false;
-    block.fChecked = false;
+    // block.m_checked_witness_commitment = false;
+    // block.m_checked_merkle_root = false;
+    // block.fChecked = false;
 }
```

ryanofsky reviewed (approved):
Code review ACK 7d57c4e44451ce1167ba76a73c9cc6f6fd6cfca4. Code, test, documentation updates all seem like improvements and it is good to make submitSolution provide better feedback. I do think more improvements could be made here: documentation could be updated to say what requirements on submitSolution inputs are, and I think BlockTemplate accessors could be added so miners do not need to make hardcoded assumptions about the coinbase transaction? (https://github.com/bitcoin/bitcoin/pull/33745#discussion_r2477980297). It would also be nice to return error messages and not just the false return value.

davidgumberg commented on test/functional/interface_ipc.py:
```diff
@@ -204,13 +208,32 @@ async def async_routine():
+
+            self.log.debug("Submit again, with the witness")
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize())
             assert_equal(res.result, True)
```
I want to suggest adding a check here:

```suggestion
            assert_equal(res.result, True)
            # self.nodes[1]'s chaintip did not advance.
            assert_equal(self.nodes[1].getchaintips()[0]["height"], current_block_height)
```

But I'm a bit confused, because testing this without the changes to `src/node/miner.cpp`, self.nodes[1]'s chaintip advances after submitSolution is called with the witnessless coinbase, and this check I'm suggesting fails. 

To reproduce, I'm using the following diff on this branch: 

```diff
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index b988e28a3f..485984e437 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -454,9 +454,9 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
     block.hashMerkleRoot = BlockMerkleRoot(block);

     // Reset cached checks
-    block.m_checked_witness_commitment = false;
-    block.m_checked_merkle_root = false;
-    block.fChecked = false;
+    // block.m_checked_witness_commitment = false;
+    // block.m_checked_merkle_root = false;
+    // block.fChecked = false;
 }

 std::unique_ptr<CBlockTemplate> WaitAndCreateNewBlock(ChainstateManager& chainman,
diff --git a/test/functional/interface_ipc.py b/test/functional/interface_ipc.py
index a6010c2fb3..0d99a6b2c3 100755
--- a/test/functional/interface_ipc.py
+++ b/test/functional/interface_ipc.py
@@ -219,7 +219,9 @@ class IPCInterfaceTest(BitcoinTestFramework):
             self.log.debug("Submitted coinbase must include witness")
             assert_not_equal(coinbase.serialize_without_witness().hex(), coinbase.serialize().hex())
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize_without_witness())
-            assert_equal(res.result, False)
+            #assert_equal(res.result, False)
+            # self.nodes[1]'s chaintip did not advance.
+            assert_equal(self.nodes[1].getchaintips()[0]["height"], current_block_height)

             self.log.debug("Submit again, with the witness")
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize())
```

And the check fails, because self.nodes[1]'s chaintip was advanced!

Sjors replied:
Added a patch to the PR description which reproduces the issue.

sedited reviewed (approved):
ACK 7d57c4e44451ce1167ba76a73c9cc6f6fd6cfca4

Sjors committed:
doc: clarify UpdateUncommittedBlockStructures

Sjors commented:
Pushed [7d57c4e](https://github.com/bitcoin/bitcoin/commit/7d57c4e44451ce1167ba76a73c9cc6f6fd6cfca4) to [2fc95ca](https://github.com/bitcoin/bitcoin/commit/2fc95ca731b52ca419215109c568e89bf683a94b): [compare](https://github.com/bitcoin/bitcoin/compare/7d57c4e44451ce1167ba76a73c9cc6f6fd6cfca4..2fc95ca731b52ca419215109c568e89bf683a94b). No behavior change.

I added a patch the to PR description that reproduces the original issue.

Expanded the `submitSolution` documentation to indicate that a duplicate block is not considered a failure, and why.

I kept all three reset caches for simplicity.

Note that `m_checked_merkle_root` will be false anyway, because `BlockAssembler::CreateNewBlock` skips the merle root check. If we really wanted to skip that check, since we created the merle root ourselves, we would have to modify `ProcessNewBlock` to forward `check_merkle_root` to its `CheckBlock` call. That seems out of scope.

Similarly `fChecked` will already be `false` because the only place that sets it to true is guarded by `if (fCheckPOW && fCheckMerkleRoot)`.

Sjors commented on src/node/miner.cpp:
```diff
@@ -452,6 +452,11 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
+
+    // Reset cached checks
+    block.m_checked_witness_commitment = false;
+    block.m_checked_merkle_root = false;
```
I kept all three, but as I explained in https://github.com/bitcoin/bitcoin/pull/33745#issuecomment-3471963619 two of them are already `false`.

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -204,13 +208,34 @@ async def async_routine():
+            assert_equal(self.nodes[1].getchaintips()[0]["height"], current_block_height + 1)
+            # Stalls if a regression causes submitBlock() to accept an invalid block:
+            self.sync_all()
+            assert_equal(self.nodes[0].getchaintips()[0], self.nodes[1].getchaintips()[0])
```
If you apply the patch in the description and also drop the `sync_all` call, you'll see that the second node sees the block as `headers-only`.

Sjors committed:
mining: ensure witness commitment check in submitBlock

When an IPC client requests a new block template via the Mining interface,
we hold on to its CBlock. That way when they call submitSolution() we can
modify it in place, rather than having to reconstruct the full block like
the submitblock RPC does.

Before this commit however we forgot to invalidate
m_checked_witness_commitment, which we should since the client brings a
new coinbase.

This would cause us to accept an invalid chaintip.

Fix this and add a test to confirm that we now reject such a block.
As a sanity check, we add a second node to the test and confirm that will
accept our mined block.

Note that the IPC code takes the coinbase as provided, unlike the
submitblock RPC which calls UpdateUncommittedBlockStructures() and adds
witness commitment to the coinbase if it was missing.

Although that could have been an alternative fix, we instead document that
IPC clients are expected to provide the full coinbase including witness
commitment.

Sjors commented:
I split the template block mutation test into a third commit and documented its history in the commit message.

Sjors committed:
test: clarify submitBlock() mutates the template

PR #33374 proposed a new Mining IPC method applySolution() which
could be used by clients to obtain the reconstructed block for
inspection, especially in the case of a rejected block.

However it was pointed out during review that submitBlock() modified
the template CBlock in place, so the client can just call getBlock()
and no new method is needed.

This commit adds a test to document that (now intentional) behavior.

ismaelsadeeq reviewed (commented):
Concept ACK, will review.

sedited reviewed (approved):
Re-ACK 6eaa00fe20206baedc0d8ade5bb8d066ea615704

ismaelsadeeq reviewed (commented):
Code review and tested ACK   6eaa00fe20206baedc0d8ade5bb8d066ea615704

This looks really good. I have verified that previously, we would indeed accept a block with an invalid witness commitment because we did not check for it. The new toggle of the cached flags to `false` in the first commit prevents us from doing so.
The added documentation also looks good.

Before the patch in the first commit, the added test in the second commit failed
```terminal
2025-11-10T17:14:58.517226Z TestFramework (INFO): PRNG seed is: 6082034237012765229
2025-11-10T17:14:58.517665Z TestFramework (INFO): Initializing test directory /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_fm1o0xxb
2025-11-10T17:15:00.198939Z TestFramework (INFO): Running echo test
2025-11-10T17:15:00.205953Z TestFramework (INFO): Running mining test
2025-11-10T17:15:06.280650Z TestFramework (ERROR): Unexpected exception
Traceback (most recent call last):
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin-work-tree/bitcoin-master/test/functional/test_framework/test_framework.py", line 142, in main
    self.run_test()
    ~~~~~~~~~~~~~^^
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin-work-tree/bitcoin-master/build/test/functional/interface_ipc.py", line 252, in run_test
    self.run_mining_test()
    ~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin-work-tree/bitcoin-master/build/test/functional/interface_ipc.py", line 248, in run_mining_test
    asyncio.run(capnp.run(async_routine()))
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.7/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/runners.py", line 195, in run
    return runner.run(main)
           ~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.7/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.7/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/base_events.py", line 725, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "capnp/lib/capnp.pyx", line 1965, in run
  File "capnp/lib/capnp.pyx", line 1966, in capnp.lib.capnp.run
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin-work-tree/bitcoin-master/build/test/functional/interface_ipc.py", line 221, in async_routine
    assert_equal(res.result, False)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin-work-tree/bitcoin-master/test/functional/test_framework/util.py", line 80, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(True == False)
2025-11-10T17:15:06.340210Z TestFramework (INFO): Not stopping nodes as test failed. The dangling processes will be cleaned up later.
2025-11-10T17:15:06.340318Z TestFramework (WARNING): Not cleaning up dir /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_fm1o0xxb
2025-11-10T17:15:06.340356Z TestFramework (ERROR): Test failed. Test logging available at /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_fm1o0xxb/test_framework.log
2025-11-10T17:15:06.340441Z TestFramework (ERROR): 
2025-11-10T17:15:06.340567Z TestFramework (ERROR): Hint: Call /Users/abubakarismail/Desktop/Work/bitcoin-dev/bitcoin-work-tree/bitcoin-master/test/functional/combine_logs.py '/var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_fm1o0xxb' to consolidate all logs
2025-11-10T17:15:06.340599Z TestFramework (ERROR): 
2025-11-10T17:15:06.340621Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
2025-11-10T17:15:06.340669Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
2025-11-10T17:15:06.340712Z TestFramework (ERROR): 
[node 1] Cleaning up leftover process
[node 0] Cleaning up leftover process
```

After compiling with the first commit, it succeeded.

```
2025-11-10T17:21:00.220383Z TestFramework (INFO): PRNG seed is: 2383671957558385254
2025-11-10T17:21:00.220788Z TestFramework (INFO): Initializing test directory /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_q5ioic08
2025-11-10T17:21:01.905456Z TestFramework (INFO): Running echo test
2025-11-10T17:21:01.912516Z TestFramework (INFO): Running mining test
2025-11-10T17:21:08.054949Z TestFramework (INFO): Stopping nodes
2025-11-10T17:21:08.163859Z TestFramework (INFO): Cleaning up /var/folders/dj/d8p8jhd172n7wnq81ryfl6rc0000gn/T/bitcoin_func_test_q5ioic08 on exit
2025-11-10T17:21:08.163922Z TestFramework (INFO): Tests successful
```
